pkg/controller/clusters/controlplane.go

```diff
@@ -44,91 +49,180 @@ type ControlPlane struct {
+	// Before making any changes to etcd, pause the KCP so it doesn't rollout new nodes as the
+	// etcd endpoints change.
+	if !annotations.HasPaused(currentKCP) {
+		log.Info("Pausing KCP before making any etcd changes", "kcp", klog.KObj(currentKCP))
+		clientutil.AddAnnotation(currentKCP, clusterv1.PausedAnnotation, "true")
+		if err := c.Update(ctx, currentKCP); err != nil {
+			return controller.Result{}, err
+		}
+	}
+
+	// If the etcdadm cluster has changes, this will require a rolling upgrade
+	// Mark the etcdadm cluster as upgrading
+	// The etcdadm cluster controller will take care of removing
+	// this annotation at the right time to orchestrate the kcp upgrade.
+	clientutil.AddAnnotation(desiredCP.EtcdCluster, etcdv1.UpgradeInProgressAnnotation, "true")
+
+	log.Info("Reconciling external etcd changes", "etcdadmCluster", klog.KObj(currentEtcdadmCluster))
+	if err := serverside.ReconcileObjects(ctx, c, desiredCP.etcdObjects()); err != nil {
+		return controller.Result{}, errors.Wrap(err, "applying etcd objects")
+	}
+
+	// After applying etcd changes, just requeue to wait until etcd finishes updating and is ready
+	// We use a short wait here just in case etcdadm controller decides that not new machines are
+	// needed.
+	log.Info("Requeuing to wait until etcd is ready")
+	return controller.ResultWithRequeue(10 * time.Second), nil
+}
+
+func etcdadmClusterReady(etcdadmCluster *etcdv1.EtcdadmCluster) bool {
+	// It's important to use status.Ready and not the Ready condition, since the Ready condition
+	// only becomes true after the old etcd members have been deleted, which only happens after the
+	// kcp finishes its own upgrade.
+	return etcdadmCluster.Generation == etcdadmCluster.Status.ObservedGeneration && etcdadmCluster.Status.Ready
+}
+
+func reconcileControlPlaneNodeChanges(ctx context.Context, log logr.Logger, c client.Client, desiredCP *ControlPlane, currentKCP *controlplanev1.KubeadmControlPlane) (controller.Result, error) {
 	// When the controller reconciles the control plane for a cluster with an external etcd configuration
 	// the KubeadmControlPlane.Spec.KubeadmConfigSpec.ClusterConfiguration.Etcd.External.Endpoints field is
 	// defaulted to an empty slice. However, at some point that field in KubeadmControlPlane object is filled
-	// and updated by another component
-
+	// and updated by the kcp controller.
+	//
 	// We do not want to update the field with an empty slice again, so here we check if the endpoints for the
 	// external etcd have already been populated on the KubeadmControlPlane object and override ours before applying it.
-	kcp := &controlplanev1.KubeadmControlPlane{}
-	kcpKey := client.ObjectKey{
-		Name:      cluster.Spec.ControlPlaneRef.Name,
-		Namespace: cluster.Spec.ControlPlaneRef.Namespace,
+	externalEndpoints := currentKCP.Spec.KubeadmConfigSpec.ClusterConfiguration.Etcd.External.Endpoints
```

Comment:
I get that ectdadm controller will remove etcdv1.UpgradeInProgressAnnotation annotation on the etcd cluster after it has been upgraded successfully but just to double check should we check the annotation doesn't exists when we upgrade the control plane?

Reply:
Ah interesting. Why do you think that could be necessary? The status contains the same information. The annotation is only there so the KCP/capi cluster controller can read it without knowing the whole etcdadmcluster API. But we do know the full api, so we don't need the validation.

Reply:
I was just thinking that if we are adding the annotation during etcd upgrade, should we check that the annotation doesn't exist before we start the CP reconciliation. I got what you are saying. Thanks!